x/fswap/keeper/keeper.go

```diff
@@ -1,42 +1,341 @@
+	oldDenom, err := k.getFromDenom(ctx)
+	if err != nil {
+		return err
+	}
+	balance := k.GetBalance(ctx, addr, oldDenom)
+	if balance.IsZero() {
+		return sdkerrors.ErrInsufficientFunds
+	}
+
+	if err := k.Swap(ctx, addr, balance); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) setSwapInit(ctx sdk.Context, swapInit types.SwapInit) error {
+	store := ctx.KVStore(k.storeKey)
+	bz, err := k.cdc.Marshal(&swapInit)
+	if err != nil {
+		return err
+	}
+	store.Set(swapInitKey(swapInit.ToDenom), bz)
+	return nil
+}
+
+func (k Keeper) getAllSwapped(ctx sdk.Context) []types.Swapped {
+	swappedSlice := make([]types.Swapped, 0)
+	k.iterateAllSwapped(ctx, func(swapped types.Swapped) bool {
+		swappedSlice = append(swappedSlice, swapped)
+		return false
+	})
+	return swappedSlice
+}
+
+func (k Keeper) iterateAllSwapped(ctx sdk.Context, cb func(swapped types.Swapped) (stop bool)) {
+	store := ctx.KVStore(k.storeKey)
+	swappedDataStore := prefix.NewStore(store, swappedKeyPrefix)
+
+	iterator := swappedDataStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		swapped := types.Swapped{}
+		k.cdc.MustUnmarshal(iterator.Value(), &swapped)
+		if cb(swapped) {
+			break
+		}
+	}
+}
+
+func (k Keeper) getSwapped(ctx sdk.Context) (types.Swapped, error) {
+	newDenom, err := k.getToDenom(ctx)
+	if err != nil {
+		return types.Swapped{}, err
+	}
+	store := ctx.KVStore(k.storeKey)
+	key := swappedKey(newDenom)
+	bz := store.Get(key)
+	if bz == nil {
+		return types.Swapped{}, types.ErrSwappedNotFound
+	}
+	swapped := types.Swapped{}
+	if err := k.cdc.Unmarshal(bz, &swapped); err != nil {
+		return types.Swapped{}, err
+	}
+	return swapped, nil
+}
+
+func (k Keeper) setSwapped(ctx sdk.Context, swapped types.Swapped) error {
+	store := ctx.KVStore(k.storeKey)
+	key := swappedKey(swapped.ToCoinAmount.Denom)
+	bz, err := k.cdc.Marshal(&swapped)
+	if err != nil {
+		return err
+	}
+	store.Set(key, bz)
+	return nil
+}
+
+func (k Keeper) getAllSwapInits(ctx sdk.Context) []types.SwapInit {
+	swapInits := make([]types.SwapInit, 0)
+	k.iterateAllSwapInits(ctx, func(swapInit types.SwapInit) bool {
+		swapInits = append(swapInits, swapInit)
+		return false
+	})
+	return swapInits
+}
+
+func (k Keeper) iterateAllSwapInits(ctx sdk.Context, cb func(swapped types.SwapInit) (stop bool)) {
+	store := ctx.KVStore(k.storeKey)
+	swapInitDataStore := prefix.NewStore(store, swapInitPrefix)
+
+	iterator := swapInitDataStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		swapInit := types.SwapInit{}
+		k.cdc.MustUnmarshal(iterator.Value(), &swapInit)
+		if cb(swapInit) {
+			break
+		}
+	}
+}
+
+func (k Keeper) getSwappableNewCoinAmount(ctx sdk.Context) (sdk.Coin, error) {
+	swapCap, err := k.getSwapCap(ctx)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	swapped, err := k.getSwapped(ctx)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	denom, err := k.getToDenom(ctx)
```

Comment:
In `MsgSwap`, the users specify the amount in "from-denom". It would be nice to provide remaining amount in "from-denom".

Reply:
I thought it's nice to provide remaining amount too. 
But I didn't add, after I checked that `Bank.Send` doesn't return remaining amount after `Send`.

Reply:
> I checked that `Bank.Send` doesn't return remaining amount after `Send`.

The bank send does not return remaining amount in the response. And I don't want `MsgSwap` to return the remaining amount either.
My point was, the query returns remaining amount in "to-denom", but in the txs we use amounts in "from-denom". But it seems the behavior is from the requirement. OK to ignore it.

Reply:
How about this?
I've added `to_denom` to `MsgSwap` and `from/to_denom` to `MsgSwapAll`.

```protobuf
service Msg {
  rpc Swap(MsgSwap) returns (MsgSwapResponse);
  rpc SwapAll(MsgSwapAll) returns (MsgSwapAllResponse);
}

message MsgSwap {
  // holder's address
  string from_address = 1;
  // from-coin amount
  cosmos.base.v1beta1.Coin from_coin_amount = 2
      [(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/Finschia/finschia-sdk/types.Coin"];
  string to_denom = 3;
}

message MsgSwapResponse {}

message MsgSwapAll {
  // holder's address
  string from_address = 1;
  string from_denom = 2;
  string to_denom = 3;
}

message MsgSwapAllResponse {}
```